client/internal/httpsender_test.go

```diff
@@ -208,3 +208,123 @@ func TestRequestInstanceUidFlagReset(t *testing.T) {
 	assert.Equal(t, sender.receiveProcessor.clientSyncedState.flags, protobufs.AgentToServerFlags_AgentToServerFlags_Unspecified)
 	cancel()
 }
+
+func TestPackageUpdatesInParallel(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	localPackageState := NewInMemPackagesStore()
+	sender := NewHTTPSender(&sharedinternal.NopLogger{})
+	ch := make(chan struct{})
+	doneCh := make([]<-chan struct{}, 0)
+
+	// Use this to simulate blocking behavior on the first call to Sync().
+	localPackageState.onAllPackagesHash = func() {
+		if localPackageState.lastReportedStatuses != nil {
+			<-ch
+		}
+	}
+
+	var messages atomic.Int32
+	var mux sync.Mutex
+	sender.callbacks = types.CallbacksStruct{
+		OnMessageFunc: func(ctx context.Context, msg *types.MessageData) {
+			err := msg.PackageSyncer.Sync(ctx)
+			assert.NoError(t, err)
+			messages.Add(1)
+			doneCh = append(doneCh, msg.PackageSyncer.Done())
```

Comment:
Can you please add a bit more comments to explain what the channels do here. I am not sure I understand the interplay between the 3 channels.

Reply:
Hey, apologies Tigran, I was a little swamped.

I've added a few comments to explain the logic in more detail, check it out: https://github.com/open-telemetry/opamp-go/pull/289/commits/76e13586a8a9b98912aaaf14bb7ccd2708a585c8

In a previous comment you said
> Is there a way we can mock InMemPackageStore to block on some operation so that we can wait for the second Sync operation to be called here in the test

Reply:
Hey, apologies Tigran, I was a little swamped.

I've added a few comments to explain the logic in more detail, check it out: https://github.com/open-telemetry/opamp-go/pull/289/commits/76e13586a8a9b98912aaaf14bb7ccd2708a585c8

In a previous comment you said
> Is there a way we can mock InMemPackageStore to block on some operation so that we can wait for the second Sync operation to be called here in the test

So that's what I tried to do. It blocks on an operation in our codepath (the `onAllPackagesHash` call), ensures that both requests are inflight at the same time, so if there was no proper mutex locking, the race detector would _always_ catch the erroneous access.